text/0000-regex-1.0.md

```diff
@@ -0,0 +1,984 @@
+  ///
+  /// `replace` replaces the first match only. `replace_all` replaces all
+  /// matches. `replacen` replaces at most `limit` matches.
+  fn replace<'t, R: Replacer>(&self, text: &'t str, rep: R) -> Cow<'t, str>;
+  fn replace_all<'t, R: Replacer>(&self, text: &'t str, rep: R) -> Cow<'t, str>;
+  fn replacen<'t, R: Replacer>(&self, text: &'t str, limit: usize, rep: R) -> Cow<'t, str>;
+}
+```
+
+And lastly, three simple accessors:
+
+```rust
+impl Regex {
+  /// Returns the original pattern string.
+  pub fn as_str(&self) -> &str;
+
+  /// Returns an iterator over all capturing group in the pattern in the order
+  /// they were defined (by position of the leftmost parenthesis). The name of
+  /// the group is yielded if it has a name, otherwise None is yielded.
+  pub fn capture_names(&self) -> CaptureNamesIter;
+
+  /// Returns the total number of capturing groups in the pattern. This
+  /// includes the implicit capturing group corresponding to the entire
+  /// pattern.
+  pub fn captures_len(&self) -> usize;
+}
+```
+
+Finally, `Regex` impls the `Send`, `Sync`, `Display`, `Debug`, `Clone` and
+`FromStr` traits from the standard library.
+
+## Error
+
+The `Error` enum is an *extensible* enum, similar to `std::io::Error`,
+corresponding to the different ways that regex compilation can fail. In
+particular, this means that adding a new variant to this enum is not a breaking
+change. (Removing or changing an existing variant is still a breaking change.)
+
+```rust
+pub enum Error {
+  /// A syntax error.
+  Syntax(SyntaxError),
+  /// The compiled program exceeded the set size limit.
+  /// The argument is the size limit imposed.
+  CompiledTooBig(usize),
+  /// Hints that destructuring should not be exhaustive.
+  ///
+  /// This enum may grow additional variants, so this makes sure clients
+  /// don't count on exhaustive matching. (Otherwise, adding a new variant
+  /// could break existing code.)
+  #[doc(hidden)]
+  __Nonexhaustive,
+}
+```
+
+Note that the `Syntax` variant could contain the `Error` type from the
+`regex-syntax` crate, but this couples `regex-syntax` to the public API
+of `regex`. We sidestep this hazard by defining a newtype in `regex` that
+internally wraps `regex_syntax::Error`. This also enables us to selectively
+expose more information in the future.
+
+## RegexBuilder
+[regexbuilder]: #regexbuilder
+
+In most cases, the construction of a regex is done with `Regex::new`. There are
+however some options one might want to tweak. This can be done with a
+`RegexBuilder`:
+
+```rust
+impl RegexBuilder {
+  /// Creates a new builder from the given pattern.
+  pub fn new(pattern: &str) -> RegexBuilder;
+
+  /// Compiles the pattern and all set options. If successful, a Regex is
+  /// returned. Otherwise, if compilation failed, an Error is returned.
+  ///
+  /// N.B. `RegexBuilder::new("...").compile()` is equivalent to
+  /// `Regex::new("...")`.
+  pub fn compile(self) -> Result<Regex, Error>;
```

Comment:
@alexcrichton Seems like it should be possible. Are you suggesting for consistency's sake, or is there a benefit to doing it with `&self` instead of `self`?